From 8aaccf00128c5df5248857004633a25560121a02 Mon Sep 17 00:00:00 2001
From: Huiwen He <hehuiwen@kylinos.cn>
Date: Wed, 18 Feb 2026 00:34:30 +0800
Subject: [PATCH 7/8] smb/client: use binary search for NT status to DOS
 mapping

The ntstatus_to_dos_map[] table is sorted. Replace the linear search with
binary search to improve lookup performance.

Also remove the sentinel entry as it is no longer needed with ARRAY_SIZE().

Signed-off-by: Huiwen He <hehuiwen@kylinos.cn>
Reviewed-by: ChenXiaoSong <chenxiaosong@kylinos.cn>
---
 fs/smb/client/smb1maperror.c | 26 ++++++++++++++++----------
 1 file changed, 16 insertions(+), 10 deletions(-)

diff --git a/fs/smb/client/smb1maperror.c b/fs/smb/client/smb1maperror.c
index 9a8da9ea6408..0f8e013655c8 100644
--- a/fs/smb/client/smb1maperror.c
+++ b/fs/smb/client/smb1maperror.c
@@ -32,25 +32,31 @@ static const struct smb_to_posix_error mapping_table_ERRSRV[] = {
 /*****************************************************************************
  *convert a NT status code to a dos class/code
  *****************************************************************************/
+static __always_inline int ntstatus_to_dos_cmp(const void *_key, const void *_pivot)
+{
+	__u32 key = *(__u32 *)_key;
+	const struct nt_err_code_struct *pivot = _pivot;
+
+	if (key < pivot->ntstatus)
+		return -1;
+	if (key > pivot->ntstatus)
+		return 1;
+	return 0;
+}
+
 /* NT status -> dos error map */
 static const struct nt_err_code_struct ntstatus_to_dos_map[] = {
 #include "smb1_mapping_table.c"
-	{0, 0, 0, NULL}
 };
 
 static const struct nt_err_code_struct *
 ntstatus_to_dos(__u32 ntstatus)
 {
-	int i;
-
-	/* Check nt_errstr to allow mapping of NT_STATUS_OK (0) */
-	for (i = 0; ntstatus_to_dos_map[i].nt_errstr; i++) {
-		if (ntstatus == ntstatus_to_dos_map[i].ntstatus) {
-			return &ntstatus_to_dos_map[i];
-		}
-	}
 
-	return NULL;
+	return __inline_bsearch(&ntstatus, ntstatus_to_dos_map,
+			       ARRAY_SIZE(ntstatus_to_dos_map),
+			       sizeof(struct nt_err_code_struct),
+			       ntstatus_to_dos_cmp);
 }
 
 int
-- 
2.53.0

